File: src/mitris/sim/core/simulation/Coordinator.java
package mitris.sim.core.simulation;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.logging.Logger;

import mitris.sim.core.Constants;
import mitris.sim.core.modeling.Coupled;
import mitris.sim.core.modeling.Coupling;
import mitris.sim.core.modeling.InPort;
import mitris.sim.core.modeling.OutPort;
import mitris.sim.core.modeling.api.Component;
import mitris.sim.core.modeling.api.DevsAtomic;
import mitris.sim.core.modeling.api.DevsCoupled;
import mitris.sim.core.simulation.api.DevsCoordinator;
import mitris.sim.core.simulation.api.DevsSimulator;
import mitris.sim.core.simulation.api.SimulationClock;
import mitris.sim.core.util.Util;

/**
 *
 * @author José Luis Risco Martín
 */
public class Coordinator extends AbstractSimulator implements DevsCoordinator {

	private static final Logger logger = Logger.getLogger(Coordinator.class.getName());

	protected DevsCoupled model;
	protected LinkedList<DevsSimulator> simulators = new LinkedList<>();

	public Coordinator(SimulationClock clock, Coupled model, boolean flatten) {
		super(clock);
		logger.fine("Hierarchical...\n" + Util.printCouplings(model));
		if(flatten) {
			this.model = model.flatten();
		}
		else {
			this.model = model;
		}
		// Build hierarchy
		Collection<Component> components = model.getComponents();
		for (Component component : components) {
			if (component instanceof Coupled) {
				Coordinator coordinator = new Coordinator(clock, (Coupled) component, false);
				simulators.add(coordinator);
			} else if (component instanceof DevsAtomic) {
				Simulator simulator = new Simulator(clock, (DevsAtomic) component);
				simulators.add(simulator);
			}
		}

		logger.fine("After flattening.....\n"+ Util.printCouplings(this.model));
		logger.fine(this.model.toString());
		Iterator<Component> itr  = this.model.getComponents().iterator();
		while(itr.hasNext()){
			logger.fine("Component: "+itr.next());
		}

	} 

	public void initialize() {
		logger.fine("START SIMULATION");        
		for (DevsSimulator simulator : simulators) {
			simulator.initialize();
		}
		tL = clock.getTime();
		tN = tL + ta();
	}

	public Coordinator(Coupled model, boolean flatten) {
		this(new SimulationClock(), model, flatten);
	}

	public Coordinator(Coupled model) {
		this(model, true);
	}

	public Collection<DevsSimulator> getSimulators() {
		return simulators;
	}

	public final double ta() {
		double tn = Constants.INFINITY;
		for (DevsSimulator simulator : simulators) {
			if (simulator.getTN() < tn) {
				tn = simulator.getTN();
			}
		}
		return tn - clock.getTime();
	}

	public void lambda() {
		for (DevsSimulator simulator : simulators) {
			simulator.lambda();
		}
		propagateOutput();
	}

	public void propagateOutput() {
		LinkedList<Coupling<?>> ic = model.getIC();
		for (Coupling<?> c : ic) {
			c.propagateValues();
		}

		LinkedList<Coupling<?>> eoc = model.getEOC();
		for (Coupling<?> c : eoc) {
			c.propagateValues();
		}
	}

	@Override
	public void deltfcn() {
		propagateInput();
		for (DevsSimulator simulator : simulators) {
			simulator.deltfcn();
		}
		tL = clock.getTime();
		tN = tL + ta();
	}

	public void propagateInput() {
		LinkedList<Coupling<?>> eic = model.getEIC();
		for (Coupling<?> c : eic) {
			c.propagateValues();
		}
	}

	public void clear() {
		for (DevsSimulator simulator : simulators) {
			simulator.clear();
		}
		Collection<InPort<?>> inPorts;
		inPorts = model.getInPorts();
		for (InPort<?> port : inPorts) {
			port.clear();
		}
		Collection<OutPort<?>> outPorts;
		outPorts = model.getOutPorts();
		for (OutPort<?> port : outPorts) {
			port.clear();
		}
	} 

	public void simulate(long numIterations) {
		clock.setTime(tN);
		long counter;
		for (counter = 1; counter < numIterations && clock.getTime() < Constants.INFINITY; counter++) {
			lambda();
			deltfcn();
			clear();
			clock.setTime(tN);
		}
	}

	public void simulate(double timeInterval) {
		clock.setTime(tN);
		double tF = clock.getTime() + timeInterval;
		while (clock.getTime() < Constants.INFINITY && clock.getTime() < tF) {
			lambda();
			deltfcn();
			clear();
			clock.setTime(tN);
		}
	}

	public DevsCoupled getModel() {
		return model;
	}
}

File: simulation/coordinator.go
package simulation

import (
	"fmt"
	"github.com/pointlesssoft/godevs/modeling"
	"github.com/pointlesssoft/godevs/util"
)

type Coordinator interface {
	AbstractSimulator
	buildHierarchy()
	GetSimulators() []*AbstractSimulator
	PropagateOutput()
	PropagateInput()
	SimInject(e float64, port *modeling.Port, values interface{})
	SimulateIterations(numIterations int)
	SimulateTime(timeInterval float64)
}

func NewCoordinator(clock *SimulationClock, model *modeling.Coupled) Coordinator {
	c := coordinator{NewAbstractSimulator(clock), model, nil, 0}
	return &c
}

type coordinator struct {
	AbstractSimulator
	model *modeling.Coupled
	simulators []*AbstractSimulator
	totalIterations int
}

func (c *coordinator) buildHierarchy() {
	panic("implement me")
}

func (c *coordinator) Initialize() {
	c.buildHierarchy()
	for _, sim := range c.simulators {
		(*sim).Initialize()
	}
	c.SetTL((*c.GetClock()).GetTime())
	c.SetTN(c.GetTL() + c.TA())
}

func (c *coordinator) Exit() {
	for _, sim := range c.simulators {
		(*sim).Exit()
	}
}

func (c *coordinator) GetSimulators() []*AbstractSimulator {
	return c.simulators
}

func (c *coordinator) TA() float64 {
	tn := util.INFINITY
	for _, sim := range c.simulators {
		simTn := (*sim).GetTN()
		if  simTn < tn {
			tn = simTn
		}
	}
	return tn - (*c.GetClock()).GetTime()
}

func (c *coordinator) Lambda() {
	for _, sim := range c.simulators {  // TODO parallelize
		(*sim).Lambda()
		c.PropagateOutput()
	}
}

func (c *coordinator) PropagateOutput() {  // TODO parallelize
	for _, coup := range (*c.model).GetIC() {
		(*coup).PropagateValues()
	}
	for _, coup := range (*c.model).GetEOC() {
		(*coup).PropagateValues()
	}
}

func (c *coordinator) DeltFcn() {  // TODO parallelize
	c.PropagateInput()
	for _, sim := range c.simulators {
		(*sim).DeltFcn()
	}
	c.SetTL((*c.GetClock()).GetTime())
	c.SetTN(c.GetTL() + c.TA())
}

func (c *coordinator) PropagateInput() {  // TODO parallelize
	for _, coup := range (*c.model).GetEIC() {
		(*coup).PropagateValues()
	}
}

func (c *coordinator) Clear() {
	for _, sim := range c.simulators {
		(*sim).Clear()
	}
	for _, port := range (*c.model).GetInPorts() {
		(*port).Clear()
	}
	for _, port := range (*c.model).GetOutPorts() {
		(*port).Clear()
	}
}

func (c *coordinator) SimInject(e float64, port *modeling.Port, values interface{}) {
	time := (*c.GetClock()).GetTime() + e
	if time <= c.GetTN() {
		(*port).AddValues(values)
		(*c.GetClock()).SetTime(time)
		c.DeltFcn()
		c.Clear()
	} else {
		panic(fmt.Sprintf("Time: %v - ERROR input rejected: elapsed time %v is not in bounds.", c.GetTL(), e))
	}
}

func (c *coordinator) SimulateIterations(numIterations int) {
	for numIterations > 0 && c.GetTN() < util.INFINITY {
		(*c.GetClock()).SetTime(c.GetTN())
		c.Lambda()
		c.DeltFcn()
		c.Clear()
		c.totalIterations++
		numIterations--
	}
}

func (c *coordinator) SimulateTime(timeInterval float64) {
	tF := (*c.GetClock()).GetTime() + timeInterval
	for (*c.GetClock()).GetTime() < util.INFINITY && c.GetTN() < tF {
		(*c.GetClock()).SetTime(c.GetTN())
		c.Lambda()
		c.DeltFcn()
		c.Clear()
	}
	(*c.GetClock()).SetTime(tF)
}

func (c *coordinator) GetModel() *modeling.Component {
	return (*c.model).GetComponent()
}
